Add note revisions

karolsojko opened the pull request:
Whenever an item is saved it will save a new revision of the item.

2 new API endpoints are created for retrieving revisions:

```
GET /items/:item_id/revisions
GET /items/:item_id/revisions/:id
```

Items to Revisions relation is many to many for the purpose of creating duplicates of items and assigning existing revisions to them.

I've also updated the local `docker-compose` setup by:
- mounting the local source code as a volume to the app container (changes in code will reflect in the running container)
- mounting `data/` folder for persisting database entries between container runs

mobitar commented on app/controllers/api/revisions_controller.rb:
```diff
@@ -0,0 +1,13 @@
+  end
+
+  def index
+    item = current_user.items.where(uuid: params[:item_id]).first
```
Can be simplified to `current_user.items.find(params[:item_id])`

mobitar commented on app/controllers/api/revisions_controller.rb:
```diff
@@ -0,0 +1,13 @@
+  def index
+    item = current_user.items.where(uuid: params[:item_id]).first
+
+    render json: item.get_revision_history(User::REVISIONS_RETENTION_DAYS)
```
`item` can perhaps be nil. So we can do something like `render json: item&.get_revision_history`

mobitar commented:
We'd also need to update the readme with API documentation so clients know how to access this feature.

johnny243 commented on spec/controllers/api/items_controller_spec.rb:
```diff
@@ -112,6 +112,29 @@
+            item = Item.where(uuid: items_param[0]['uuid']).first
+
+            revisions = item.revisions
+            expect(revisions.count).to equal(3)
```
This is failing for me. It is expected to be 3 but got 2 instead.

karolsojko replied:
Maybe the 2 consecutive POST requests are throttled by the sync manager? Meaning they happen too fast one after another. Maybe adding a `sleep` line between those codes make sense - @mobitar wdyt?

mobitar replied:
Hmm as far as I know these requests should be synchronous.

karolsojko replied:
works on my local setup and CI tests are successfull as well. Are we investigating this further?

Btw. the count should be 3 (1 revision when creating, 2 after udpates)

johnny243 replied:
I think we should investigate this, I ran the tests multiple times and it seems this particular test is failing randomly. Here's the output from `rspec`:

![image](https://user-images.githubusercontent.com/5891646/83642398-c9e55880-a57c-11ea-82f2-d0282db957d0.png)

karolsojko replied:
🤔 will give this a look

karolsojko replied:
Found the reason for this.

The Sync Manager is discarding requests that are 1 second apart from what it already has. In the test I was making the same request twice. So what I did to fix this was to make the second request with the item that I retrieved as a response from the first request.

source: https://github.com/standardnotes/syncing-server/blob/develop/lib/sync_engine/2019_05_20/sync_manager.rb#L78-L90

karolsojko commented:
> We'd also need to update the readme with API documentation so clients know how to access this feature.

@mobitar we currently don't have any API description in the README ... maybe it makes more sense to tackle this in a separate issue. Something along the lines of "preparing live API docs" - maybe with Swagger ?

mobitar commented:
True, I suppose the closest we have to API documentation is this: https://docs.standardnotes.org/specification/sync

But hasn't been updated in some time. How do you propose communicating to client developers these immediate changes? (Not just internal developers but external as well) For example the new Sessions API is documented here: https://github.com/standardnotes/syncing-server/blob/004-new/SESSIONS.md

karolsojko commented:
> True, I suppose the closest we have to API documentation is this: https://docs.standardnotes.org/specification/sync
> 
> But hasn't been updated in some time. How do you propose communicating to client developers these immediate changes? (Not just internal developers but external as well) For example the new Sessions API is documented here: https://github.com/standardnotes/syncing-server/blob/004-new/SESSIONS.md

I think we should create a ticket in Asana to update these docs that are outdated. Then pasting the link to the documentation in the README - wdyt?

mobitar commented:
Sounds good. I like the idea of an automated solution 👍 We should perhaps get to this sooner than later.

karolsojko commented:
@mobitar @johnny243 

I've fixed a couple of things:
- the revision history is cleaned up on item's `before_save` callback. Algorithm taken from `note-history-ext` with a little modification of destroying_all records that do not meet criteria.
- added index for destroy purposes 
- added cascading destroy on itemRevisions when a revision is cleaned up
- Added a 404 spec for revisions controller 
- Fixed the spec for returning limited amount of revisions from 30 days [should be 151 revisions = 30 days * 5 revisions (every 2nd revsion returned, so indexes: 0,2,4,6,8) + last day has an extra sync so that's a +1]

mobitar commented on spec/controllers/api/items_controller_spec.rb:
```diff
@@ -112,6 +112,29 @@
+            item = Item.where(uuid: items_param[0]['uuid']).first
+
+            revisions = item.revisions
+            expect(revisions.count).to eq(3)
```
Wait why do we expect 3 and not 2? I'm only seeing 2 sync requests made 🤔

johnny243 replied:
It should be 1 revision for when the item is created, and 2 revisions for the sync requests above.

karolsojko replied:
1st revision is made upon the creation of an item, 2 following ones are the updates via sync

mobitar commented on app/models/item.rb:
```diff
@@ -59,4 +67,45 @@ def perform_associated_job
+
+  def cleanup_excessive_revisions(days = User::REVISIONS_RETENTION_DAYS)
+    days.times do |days_from_today|
+      allowed_revisions_count = [days - days_from_today + MIN_REVISIONS_PER_DAY, MAX_REVISIONS_PER_DAY].min
```
I'm not sure I'm following this. So here's how the current ext works:

Today (0 days from present): 30 revisions
Today - 1 (yesterday) (1 days from present): 29 revisions
Today - 2 (2 days from present): 28 revisions
...
Today - 40 (40 days from present): min_revisions : 2

Whereas with this current functionality, it seems we have:

Today (0 days from present): `[30 - 0 + MIN_REVISIONS_PER_DAY, MAX_REVISIONS_PER_DAY].min = [34, 8].min = 8`
Today - 1 (yesterday) (1 days from present):  `[30 - 1 + MIN_REVISIONS_PER_DAY, MAX_REVISIONS_PER_DAY].min = [33, 8].min = 8`
...

So it always returns 8?

In any case, we should also add a test that correctly anticipates the decaying revision count.

karolsojko replied:
Isn't this the same as in the extension?

https://github.com/standardnotes/note-history-ext/blob/efdb5db80ab04a1fd907d6f9be6c803a93ad3519/app/models/revisions/item.rb#L29

mobitar commented:
Oh one other note: we should only create revisions for items where item.content_type == 'Note'. While it's good we designed the system to be generic, the clients will only support Note revisions for the foreseeable future.

karolsojko commented:
Hey @mobitar @johnny243 

I've update the PR with the following:
- Fixed the race condition when it was sometimes returning 2 revisions instead of 3
- Fixed the cleanup algorithm to leave the amount of revisions in a decaying fashion as described by @mobitar 
- Added the cleanup of revisions as a `perform_later` job since it doesn't need to happen on the same request-response transaction for the user. My idea is that the instead of "immediate" we need an "eventual" cleanup, thus performance-wise it's better to handle this in a background job instead of blocking the user from getting a response.
- Added specs for checking the general count of revisions after cleanup 
- Added specs for checking the day-by-day count of revisions after cleanup 
- Fixed saving revisions only when an item is a note

karolsojko commented:
In addition to this I think we also need a scheduled cron job running once a day to either:
- cleanup revisions older than X days
- do the same cleaup algorithm on every item (performance-wise tricky)

Wdyt @mobitar ?

mobitar commented on app/controllers/api/revisions_controller.rb:
```diff
@@ -0,0 +1,17 @@
+    begin
+      item = current_user.items.find(params[:item_id])
+    rescue ActiveRecord::RecordNotFound
+      return render json: { error: 'Item not found' }, status: :not_found
```
The error format we [typically use](https://github.com/standardnotes/syncing-server/blob/develop/app/controllers/api/auth_controller.rb#L115) (and clients expect) in the API is `{error : { message: '' } }`

mobitar commented on db/migrate/20200601101128_add_item_revisions.rb:
```diff
@@ -0,0 +1,22 @@
+      t.string "uuid", limit: 36, primary_key: true, null: false
+      t.text "content", limit: 16.megabytes - 1
+      t.string "content_type"
+      t.text "enc_item_key"
```
We would need `items_key_id` here as well from the new 004-new brach. This PR should really be targeted at that branch rather than develop.

karolsojko replied:
what exactly is the `items_key_id` and how is it different to `item.uuid` ?

mobitar replied:
Recommend reading [this](https://github.com/standardnotes/snjs/blob/004/specification.md) :) items_key_id is the uuid of the items_key item used to encrypt that item.

karolsojko replied:
As discussed - we will add this once https://github.com/standardnotes/syncing-server/pull/56 gets merged 

This PR should come after it

johnny243 replied:
@karolsojko #56 has been merged :+1:

karolsojko replied:
rebased & added `items_key_id` as well to `revision`

mobitar commented on app/models/item.rb:
```diff
@@ -59,4 +63,26 @@ def perform_associated_job
+  private
+
+  def cleanup_excessive_revisions(days = User::REVISIONS_RETENTION_DAYS)
+    CleanupRevisionsJob.perform_later(uuid, days)
```
Should also have content_type == Note check?

mobitar commented on app/models/item.rb:
```diff
@@ -3,11 +3,12 @@ class Item < ApplicationRecord
   has_many :revisions, -> { order 'revisions.created_at DESC' }, through: :item_revisions, dependent: :destroy
 
-  after_commit :persist_revision, :cleanup_excessive_revisions
+  after_commit :persist_revision, :cleanup_excessive_revisions, :duplicate_revisions
```
We don't want to do the duplicate revisions job after every save do we? The duplicate_of field is permanent once set, so if you save the duplicate item many items, it will run the duplicate job many times. Perhaps take a look at `after_create` instead?

karolsojko replied:
good catch, will fix

karolsojko commented on app/jobs/cleanup_revisions_job.rb:
```diff
@@ -22,10 +22,37 @@ def cleanup_revisions_for_a_day(item, days_from_today, allowed_revisions_count)
+      beginning_counter = 0
+      end_counter = revisions_divided_into_slices.length - 1
+      counter = 0
+      while revisions_to_keep.length < allowed_revisions_count
```
Then it doesn't even make it in here since line 18 `if revisions_from_date_count > allowed_revisions_count`

mobitar replied:
Ah, yup.

mobitar commented on app/jobs/duplicate_revisions_job.rb:
```diff
@@ -0,0 +1,10 @@
+class DuplicateRevisionsJob < ApplicationJob
+  def perform(item_id)
+    item = Item.find(item_id)
+    original_item_revisions = ItemRevision.where(item_uuid: item.duplicate_of).pluck(:revision_uuid)
```
We should probably add some sort of authentication here to prove that the duplicate_of belongs to the same user. Otherwise a client can put any duplicate_of UUID and the server will obey.

mobitar commented on app/jobs/duplicate_revisions_job.rb:
```diff
@@ -1,10 +1,19 @@
+      .where(uuid: item.duplicate_of, user_uuid: item.user_uuid)
+      .first
+
+    if existing_original_item do
```
Needless `do` here?

karolsojko commented:
> This is failing for me:
> 
> ![image](https://user-images.githubusercontent.com/5891646/84723518-c0cc9200-af53-11ea-8afa-3aeb5bdc289f.png)

@johnny243 can you update the branch? Works for me and as you can see on the checks PR / test has passed
